spock-core/src/main/java/spock/lang/Snapshotter.java

```diff
@@ -0,0 +1,295 @@
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package spock.lang;
+
+import org.codehaus.groovy.runtime.StringGroovyMethods;
+import org.jetbrains.annotations.NotNull;
+import org.spockframework.runtime.Condition;
+import org.spockframework.runtime.ConditionNotSatisfiedError;
+import org.spockframework.runtime.model.FeatureInfo;
+import org.spockframework.runtime.model.IterationInfo;
+import org.spockframework.runtime.model.TextPosition;
+import org.spockframework.util.Beta;
+import org.spockframework.util.Checks;
+import org.spockframework.util.IoUtil;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * Allows to perform snapshot testing.
+ * <p>
+ * Snapshots are stored in a file next to the test class.
+ *
+ * @author Leonard Brünings
+ * @since 2.4
+ */
+@Beta
+public class Snapshotter {
+  private final Store snapshotStore;
+  private Wrapper wrapper = Wrapper.NOOP;
+
+  private Function<String, String> normalizer = StringGroovyMethods::normalize;
+
+  public Snapshotter(Store snapshotStore) {
+    this.snapshotStore = snapshotStore;
+  }
+
+  protected Store getSnapshotStore() {
+    return snapshotStore;
+  }
+
+  protected String loadSnapshot(String snapshotId) {
+    return getSnapshotStore().loadSnapshot(snapshotId).map(wrapper::unwrap).orElse("<Missing Snapshot>");
+  }
+
+  protected void saveSnapshot(String snapshotId, String value) {
+    snapshotStore.saveSnapshot(snapshotId, wrapper.wrap(value));
+  }
+
+  /**
+   * Declares a {@link Wrapper} for wrapping and unwrapping of the reference value.
+   * <P>
+   * It is suggested that users create subclasses of {@link Snapshotter} with common wrapped types,
+   * instead of using this method directly.
+   */
+  public Snapshotter wrappedAs(Wrapper wrapper) {
+    this.wrapper = wrapper;
+    return this;
+  }
+
+  /**
+   * Declares a {@link Function} for normalizing the reference value.
+   * <P>
+   * The default normalization is line ending normalization.
+   */
+  public Snapshotter normalizedWith(Function<String, String> normalizer) {
+    this.normalizer = normalizer;
+    return this;
+  }
+
+  /**
+   * Specifies the {@code actual} value.
+   * <p>
+   * Note: The value is normalized by the normalizer function, defaults to line ending normalization.
+   * @param value the actual value for assertions
+   */
+  public Snapshot assertThat(String value) {
+    return new Snapshot(value);
+  }
+
+  @Override
+  public String toString() {
+    return "Snapshotter";
+  }
+
+  public class Snapshot {
+    private final String value;
```

Comment:
```suggestion
    @Nullable
    private final String value;
```

Reply:
Why nullable? It should always have a value.

Reply:
Ok, if it is non-null, then I would expect a null check in the constructor, before and after the `normalizer`.